Source/NSTreeController.m

```diff
@@ -127,123 +155,242 @@ - (BOOL) selectsInsertedObjects
   return _selectsInsertedObjects;
 }
 
-- (BOOL) setSelectionIndexPath: (NSIndexPath*)indexPath
+- (BOOL) setSelectionIndexPath: (NSIndexPath *)indexPath
 {
-  // FIXME
-  return NO;
+  BOOL f = [self commitEditing];
+
+  if (YES == f)
+    {
+      [_selection_index_paths addObject: indexPath];
+    }
+
+  return f;
 }
 
-- (BOOL) setSelectionIndexPaths: (NSArray*)indexPaths
+- (BOOL) setSelectionIndexPaths: (NSArray *)indexPaths
 {
-  // FIXME
-  return NO;
+  BOOL f = [self commitEditing];
+
+  if (YES == f)
+    {
+      NSMutableArray *mutable_index_paths = [NSMutableArray arrayWithArray: indexPaths];
+      ASSIGN(_selection_index_paths, mutable_index_paths);
+    }
+
+  return f;
 }
 
-- (id) arrangedObjects
+- (NSTreeNode *) arrangedObjects
 {
-  // FIXME
-  return nil;
+  if (_arranged_objects == nil)
+    {
+      [self rearrangeObjects];
+    }
+  return _arranged_objects;
 }
 
-- (id) content
+- (void) rearrangeObjects
 {
-  // FIXME
-  return [super content];
+  [self willChangeValueForKey: @"arrangedObjects"];
+  DESTROY(_arranged_objects);
+
+  if ([_content isKindOfClass: [NSArray class]])
+    {
+      _arranged_objects = [[GSControllerTreeProxy alloc] initWithContent: _content
+							  withController: self];
+    }
+
+  [self didChangeValueForKey: @"arrangedObjects"];
 }
 
-- (NSArray*) selectedObjects
+- (id) _objectAtIndexPath: (NSIndexPath *)indexPath
 {
-  // FIXME
-  return [super selectedObjects];
+  NSUInteger length = [indexPath length];
+  NSUInteger pos = 0;
+  NSMutableArray *children = [_arranged_objects mutableChildNodes];
+  NSUInteger lastIndex = 0;
+  id obj = nil;
+
+  for (pos = 0; pos < length - 1; pos++)
+    {
+      NSUInteger i = [indexPath indexAtPosition: pos];
+      id node = [children objectAtIndex: i];
+
+      children = [node valueForKeyPath: _childrenKeyPath];
+    }
+
+  lastIndex = [indexPath indexAtPosition: length - 1];
+  obj = [children objectAtIndex: lastIndex];
+
+  return obj;
 }
 
-- (NSIndexPath*) selectionIndexPath
+- (NSArray *) selectedObjects
 {
-  // FIXME
-  return nil;
+  NSMutableArray *selectedObjects = [NSMutableArray array];
+
+  FOR_IN(NSIndexPath*, path, _selection_index_paths)
+    {
+      id obj = [self _objectAtIndexPath: path];
+      [selectedObjects addObject: obj];
+    }
+  END_FOR_IN(_selection_index_paths);
+
+  return selectedObjects;
 }
 
-- (NSArray*) selectionIndexPaths
+- (NSIndexPath *) selectionIndexPath
 {
-  // FIXME
-  return nil;
+  return [_selection_index_paths objectAtIndex: 0];
+}
+
+- (NSArray *) selectionIndexPaths
+{
+  return _selection_index_paths;
```

Comment:
Shouldn't we return an immutable copy here?

Reply:
Done